src/common/cockpitcloserange.c

```diff
@@ -0,0 +1,110 @@
+#include "cockpitcloserange.h"
+
+#include <assert.h>
+#include <dirent.h>
+#include <err.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/resource.h>
+
+typedef struct {
+  int from;
+  int until;
+} CloseRange;
+
+static int
+closefd (void *data,
+         int fd)
+{
+  CloseRange *ca = data;
+  if (fd >= ca->from && fd <= ca->until)
+    {
+      while (close (fd) < 0)
+        {
+          if (errno == EAGAIN || errno == EINTR)
+            continue;
+          if (errno == EBADF || errno == EINVAL)
+            break;
+          warnx ("couldn't close fd: %m");
+          return -1;
+        }
+    }
+
+  return 0;
+}
+
+#ifndef HAVE_FDWALK
+
+static int
+fdwalk (int (*cb)(void *data, int fd),
+        void *data)
+{
+  int open_max;
+  int fd;
+  int res = 0;
+
+  struct rlimit rl;
+
+#ifdef __linux__
+  DIR *d;
+
+  if ((d = opendir ("/proc/self/fd"))) {
+      struct dirent *de;
+
+      while ((de = readdir (d))) {
+          long l;
+          char *e = NULL;
+
+          if (de->d_name[0] == '.')
+              continue;
+
+          errno = 0;
+          l = strtol (de->d_name, &e, 10);
+          if (errno != 0 || !e || *e)
+              continue;
+
+          fd = (int) l;
+
+          if ((long) fd != l)
+              continue;
+
+          if (fd == dirfd (d))
+              continue;
+
+          if ((res = cb (data, fd)) != 0)
+              break;
+        }
+
+      closedir (d);
+      return res;
+  }
+
+  /* If /proc is not mounted or not accessible we fall back to the old
+   * rlimit trick */
+
+#endif
+
+  if (getrlimit (RLIMIT_NOFILE, &rl) == 0 && rl.rlim_max != RLIM_INFINITY)
+      open_max = rl.rlim_max;
+  else
+      open_max = sysconf (_SC_OPEN_MAX);
+
+  for (fd = 0; fd < open_max; fd++)
+      if ((res = cb (data, fd)) != 0)
+          break;
+
+  return res;
+}
+
+#endif /* HAVE_FDWALK */
+
+
+int
+cockpit_close_range (int from, int max_fd, int flags)
+{
+  assert (flags == 0);
```

Comment:
Just drop the flags parameter, if you're going to do that.  This is an internal API after all.

Drop the `max_fd` paramter while you're at it, since there is one user of it that should be refactored anyway.

Reply:
Nack -- the *point* was to move the `close_range()` API, which has this very API. As the next step this will call the actual `close_range()` once it's available in glibc, and keep the current code as a fallback, and eventually we can entirely drop it.